frame/support/procedural/src/pallet/expand/storage.rs

```diff
@@ -126,6 +130,31 @@ pub fn expand_storages(def: &mut Def) -> proc_macro2::TokenStream {
 							value: #frame_support::metadata::DecodeDifferent::Encode(#value),
 						}
 					)
+				},
+				Metadata::NMap { keys, hashers, value, .. } => {
+					let keys = keys
+						.iter()
+						.map(|key| clean_type_string(&quote::quote!(#key).to_string()))
+						.collect::<Vec<_>>();
+					let hashers = hashers
+						.iter()
+						.map(|hasher| syn::Ident::new(
+							&clean_type_string(&quote::quote!(#hasher).to_string()),
```

Comment:
So since `StorageHasher` is a trait, the proper way to type the associated constant is `&'static [&'static dyn StorageHasher]`, and we can't really make it as a trait object because it'll then require me to specify the concrete type for the associated type `Output` is.

Reply:
no I mean we only need the metadata of storage hasher, so frame_metadata::StorageHasher, https://crates.parity.io/frame_metadata/enum.StorageHasher.html

So something like this:
```diff
diff --git a/frame/support/src/storage/types/key.rs b/frame/support/src/storage/types/key.rs
index 9b3bee1e6f..10221562dc 100755
--- a/frame/support/src/storage/types/key.rs
+++ b/frame/support/src/storage/types/key.rs
@@ -49,12 +49,14 @@ pub trait KeyGenerator {
 		key: &KArg,
 		hash_fns: Self::HArg,
 	) -> Vec<u8>;
+	const HASHERS_METADATA: &'static [frame_metadata::StorageHasher];
 }
 
 /// A trait containing methods that are only implemented on the Key struct instead of the entire tuple.
 pub trait KeyGeneratorInner: KeyGenerator {
 	/// Hash a given `encoded` byte slice using the `KeyGenerator`'s associated `StorageHasher`.
 	fn final_hash(encoded: &[u8]) -> Vec<u8>;
+	const HASHER_METADATA: frame_metadata::StorageHasher;
 }
 
 impl<H: StorageHasher, K: FullCodec> KeyGenerator for Key<H, K> {
@@ -83,12 +85,14 @@ impl<H: StorageHasher, K: FullCodec> KeyGenerator for Key<H, K> {
 				.expect("should have at least one element!"),
 		)
 	}
+	const HASHERS_METADATA: &'static [frame_metadata::StorageHasher] = &[H::METADATA];
 }
 
 impl<H: StorageHasher, K: FullCodec> KeyGeneratorInner for Key<H, K> {
 	fn final_hash(encoded: &[u8]) -> Vec<u8> {
 		H::hash(encoded).as_ref().to_vec()
 	}
+	const HASHER_METADATA: frame_metadata::StorageHasher = H::METADATA;
 }
 
 #[impl_trait_for_tuples::impl_for_tuples(2, 18)]
@@ -125,6 +129,9 @@ impl KeyGenerator for Tuple {
 		);
 		migrated_key
 	}
+	const HASHERS_METADATA: &'static [frame_metadata::StorageHasher] = &[
+		for_tuples!( #( Tuple::HASHER_METADATA ),*  )
+	];
 }
 
 /// Marker trait to indicate that each element in the tuple encodes like the
diff --git a/frame/support/src/storage/types/nmap.rs b/frame/support/src/storage/types/nmap.rs
index 6e29c05b1d..5e9f42dbf1 100755
--- a/frame/support/src/storage/types/nmap.rs
+++ b/frame/support/src/storage/types/nmap.rs
@@ -352,6 +352,7 @@ pub trait StorageNMapMetadata {
 	const MODIFIER: StorageEntryModifier;
 	const NAME: &'static str;
 	const DEFAULT: DefaultByteGetter;
+	const HASHERS: &'static [frame_metadata::StorageHasher];
 }
 
 impl<Prefix, Key, Value, QueryKind, OnEmpty> StorageNMapMetadata
@@ -368,6 +369,7 @@ where
 	const DEFAULT: DefaultByteGetter = DefaultByteGetter(
 		&OnEmptyGetter::<QueryKind::Query, OnEmpty>(core::marker::PhantomData),
 	);
+	const HASHERS: &'static [frame_metadata::StorageHasher] = Key::HASHERS_METADATA;
 }
 
 #[cfg(test)]
```

Reply:
Oh right, `frame_metadata::StorageHasher` is an enum, not a trait like `hash::StorageHasher`. It's so easy to mess this up with so many files open!